SLE-888: Split debug logs and IDE-specific traces

thahnen opened the pull request:
## Summary

The **SonarLint Console** of the builtin Console view currently offers the user the option to display the *Verbose output* (containing the debug statements) and the *Analysis logs* next to the default output.

For the last release, I changed the behavior for Eclipse-specific invocations (e.g., adaptations or extension-point interaction) to also log their output that can be used when debugging third-party integrations with SonarLint (e.g., from IBM Developer for z/OS). These logs are very useful but pollute the output for users who want to debug "normal" issues with SonarLint.

See this exemplary screenshot for example, on how much of the output is not coming from either the *Verbose output* (actual debug logs of SonarLint logic) and *Analysis logs*. Colors are OS-specific but differ:

![Screenshot from 2024-06-20 14-15-00](https://github.com/SonarSource/sonarlint-eclipse/assets/6588381/2e284d6a-f839-4887-8bb3-ca5c7f3d5170)

In order to keep the normal *Verbose output* simple and SonarLint-logic-specific, a new option, *IDE-specific traces*, was added to the **SonarLint Console** that can be enabled separately when we ask for it. For now, the behavior for asking users for "logs" will be for them to have *Verbose output* and *Analysis logs* enabled, they are distinguished by color.
When we want to go into more detail, as this is not used in most cases, we can ask them to also enable the *IDE-specific traces* option - it is called *tracing* in context of Eclipse IDE.

## Testing

This is already used in the `StandaloneAnalysisTest#test_jdt_java_like_extension_COBOL()` but is not necessary for the other unit/integration tests. Therefore the option is disabled by default.

thahnen committed:
SLE-888: Split debug logs and IDE-specific traces

Currently the debug logs will contain all the traces we use for Eclipse IDE specific debugging, like adaptations of different elements to SonarLint-specific classes or interaction with the extension points provided by SonarLint.

This is very useful information but does collide with the debug logs in may ways and is only necessary for Eclipse-specific debugging.

sophio-japharidze-sonarsource commented on its/src/org/sonarlint/eclipse/its/StandaloneAnalysisTest.java:
```diff
@@ -137,17 +137,25 @@ public void analyze_automatic_workspace_build_disabled() {
+      var consoleText = new SonarLintConsole().getConsoleView().getConsoleText();
+      assertThat(consoleText)
+        .contains("Execute Sensor: TextAndSecretsSensor")
+        .doesNotContain("File 'Test.cbl' excluded by 'JavaProjectConfiguratorExtension'");
```
no big deal but this test does not really guarantee that the IDE specific logs will work, it checks that the log line will NOT be there - either it's really not there or the logging is not working :D 

If possible, maybe we could also assert that something from the IDE logs IS there

thahnen replied:
Good idea, I see what a reliable log message we could use

thahnen replied:
Done

sophio-japharidze-sonarsource commented on its/src/org/sonarlint/eclipse/its/reddeer/views/SonarLintConsole.java:
```diff
@@ -70,6 +70,16 @@ public void enableAnalysisLogs() {
+  public void enableIdeSpecificLogs(boolean enabled) {
+    consoleView.activate();
+    var menu = new ToolItemMenuItem(new DefaultToolItem(consoleView.getCTabItem().getFolder(), "Configure logs"), "IDE-specific traces");
+    if (!menu.isSelected() && enabled) {
```
identical implementation in if-else branches??

thahnen replied:
No, the `menu.select()` is both used for selecting and de-selecting :face_in_clouds: I can merge the branches tho and add a comment for when it applies:
- we don't want it enabled, but currently it is
- we want it enabled, and currently, it is not

thahnen replied:
Done

sophio-japharidze-sonarsource commented on org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/LogListener.java:
```diff
@@ -29,4 +29,10 @@ public interface LogListener {
+   *  ones are handled like debug messages. IDE-specific logging is something like adaptations, interaction with an
+   *  extension point, ...
+   */
+  void trace(@Nullable String msg);
```
I would suggest to rename this method to something more descriptive like `traceIdeMessage`, otherwise it is guaranteed to be confused with SLCORE trace method, even though you have nice comments on top :)

thahnen replied:
Done

sophio-japharidze-sonarsource reviewed (commented):
Please have a look at the comments and LMK in case they don't make sense to you.

thahnen committed:
SLE-888: PR feedback

Moved `trace(...)` to `traceIdeMessage(...)` in order to not confuse it with SLCORE tracing messages.

Updated the integration test to clarify that the IDE-specific logs work. Also, moved to Awaitility library.

sophio-japharidze-sonarsource commented on its/src/org/sonarlint/eclipse/its/reddeer/views/SonarLintConsole.java:
```diff
@@ -70,6 +70,18 @@ public void enableAnalysisLogs() {
+    // When the IDE-specific logs are not selected but we want to enable them, we "select" the menu item.
+    // When the IDE-specific logs are selected but we want to disable them, we "de-select" the menu item.
+    // -> `menu.select()` is used for toggling them on/off, not the greatest naming.
+    if ((!menu.isSelected() && enabled) || (menu.isSelected() && !enabled)) {
```
Is the 'if statement' even needed? maybe the comment above is enough

thahnen replied:
Yes it is, so we don't accidentily trigger the menu item on or off. As I said it can be both selecting and de-selecting.
This if statement is used so we only select it if we really want to have it enabled (if already enabled, nothing needs to be done) and we de-select it if we want to have it disabled (if already disabled, nothing needs to be done).

sophio-japharidze-sonarsource commented on org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/preferences/SonarLintPreferencesInitializer.java:
```diff
@@ -32,7 +32,9 @@ public class SonarLintPreferencesInitializer extends AbstractPreferenceInitializ
   public void initializeDefaultPreferences() {
     var node = DefaultScope.INSTANCE.getNode(SonarLintUiPlugin.PLUGIN_ID);
     node.put(SonarLintConsole.P_SHOW_CONSOLE, SonarLintConsole.P_SHOW_CONSOLE_ON_ERROR);
+    node.putBoolean(SonarLintConsole.P_VERBOSE_OUTPUT, false);
```
What was happening before this change? verbose output was enabled by default?

thahnen replied:
This is not linked to the UI. This is for the preferences saved on disk as Eclipse does preserve them this way.
E.g. when exporting the preferences of the workspace (the settings) to share them with others, this is used to fill in the default values.
When the other one was added some time ago, I assume this was just forgotten.
